### ISSUE TITLE
GitHub v4 API support

### ISSUE THREAD

**Issue by bitwiseman** [kohsuke/github-api#521]
The v4 API provides a much more customizable API based on GraphQL.  It is out of scope for this library to provide a generalized API the fully leverages the power of GraphQL, but having some way to construct queries for trees of items would be useful. 

https://developer.github.com/v4/

**Comment by bitwiseman** [comment]
https://github.com/hub4j/github-api/pull/791#issuecomment-619786105

> To connect to GH v4 API Im using "Allegro" framework (https://github.com/apollographql/apollo-android)
> 
> Calls itself are rather simple..
> Only part that Is common to Github and I hat to write small wrapper over framework was paging and rate limits..

> For now Im just using GraphQL to:
> - Setup branch protections (v4 allows setting branch protection with rules not branch per branch so In my case for some repos I was able ro replace 500 rules for 5)
> - Get list of all users with SAML emails.. to check with our AD and remove ones that are not there anymore.

**Comment by bitwiseman** [comment]
As noted in #1319 
Side note:
The objects returned as part of events are NOT the same as the GraphQL version of those same objects, Discussions in that issue. https://docs.github.com/en/graphql/reference/objects#discussion .

GraphQL `Node`s have `id` as `ID!` which is a Base64 string - https://docs.github.com/en/graphql/reference/scalars#id  but the objects returned with events return `id` as `long` and have `node_id` field for the GraphQL `id` value. 😠 And you can't mix and match - REST API requires numeric ids to reference objects, while GraphQL requires `ID!` ids.  

😡 😡 

This means that when we start implementing and moving to GraphQL there will need to be some way to convert objects returned with events to their GraphQL equivalents.

**Comment by bitwiseman** [comment]
Breaking changes do occur in GraphQL API, but they are rare:
https://docs.github.com/en/graphql/overview/breaking-changes

Non-breaking changes happen much more often: 
https://docs.github.com/en/graphql/overview/changelog

**Comment by jgangemi** [comment]
has there been any further thought about this? i have a feeling that access to dependabot alerts are only going to be available via graphql.

i guess there's nothing that would currently prevent me from using that endpoint as well...

**Comment by bitwiseman** [comment]
Yeah, there's a bunch of underlying work to be done here. But this might be a case where we start with something closer to the raw API and then make it more Java friendly over time.

**Comment by jglick** [comment]
IIUC a key advantage of GraphQL is that you can generate type-safe client interfaces from schema. That sounds like a different library altogether, though there might be a bit of overlap (or at least a factored-out common library) to deal with authentication.

**Comment by bitwiseman** [comment]
Just generating the schema is not enough. 


Authentication, rate limits, retrying, stable api that compensates for bugs and backward compatibility. 

Also GraphQL offers more data but requires the client to specify the fields they want. Which means everyone can generate queries for exactly the data they need, but if they are not sure of what data is needed at query time (as with Jenkins, with its plugins), they have to request all possible data. Which defeats one of the key features of GraphQL. 

Shrug.

**Comment by jglick** [comment]
> if they are not sure of what data is needed at query time (as with Jenkins, with its plugins)

Perhaps; depends on what code is constructing the query and for what purpose. If it is exposing some sort of GH object for other unknown plugins to inspect in arbitrary ways, then yes this would defeat a key aspect of GraphQL.

**Comment by gsmet** [comment]
FWIW, in Quarkus GitHub App, I expose both the GitHub REST API using this very API and a low level GraphQL client.
From my experience, both are useful for different things and they both have unique features.
For GraphQL, I don't think I would imagine not writing my queries myself as they really need to be tuned and limited to the strict minimum to avoid raising costs (the GraphQL API rate limiting is cost-based).

**Comment by samrocketman** [comment]
### Purpose

The purpose of using GraphQL is resolving the same information required by plugins but using GraphQL instead of the GitHub search API (which is inefficient).

You don't need a general purpose "plugins request whatever they want" kind of client.  You simply need to provide the same metadata currently provided for plugins but using GraphQL instead of the GitHub search API.

### Jervis has its own client

Jervis has its own client specifically for GraphQL because this library was not up to the task.  I would recommend either using it or relying on something similar in concept.

* Client: [`GitHubGraphQL`](https://github.com/samrocketman/jervis/blob/jervis-2.0/src/main/groovy/net/gleske/jervis/remotes/GitHubGraphQL.groovy) - [Java documentation](https://sam.gleske.net/jervis-api/2.0/net/gleske/jervis/remotes/GitHubGraphQL.html)
* Credential for this client using GitHub app with scoping available.
  * [`GitHubAppRsaCredentialImpl`](https://github.com/samrocketman/jervis/blob/jervis-2.0/src/main/groovy/net/gleske/jervis/remotes/creds/GitHubAppRsaCredentialImpl.groovy) (GH app RSA private key) - [Java documentation](https://sam.gleske.net/jervis-api/2.0/net/gleske/jervis/remotes/creds/GitHubAppRsaCredentialImpl.html)
  * [`EphemeralTokenCache`](https://github.com/samrocketman/jervis/blob/jervis-2.0/src/main/groovy/net/gleske/jervis/remotes/creds/EphemeralTokenCache.groovy) (an encrypted backend cache if the issued token is still valid or renews if it expires within a threshold) - [Java documentation](https://sam.gleske.net/jervis-api/2.0/net/gleske/jervis/remotes/creds/EphemeralTokenCache.html)
  * [`GitHubAppCredential`](https://github.com/samrocketman/jervis/blob/jervis-2.0/src/main/groovy/net/gleske/jervis/remotes/creds/GitHubAppCredential.groovy) (resolves scoped tokens for the client). - [Java documentation](https://sam.gleske.net/jervis-api/2.0/net/gleske/jervis/remotes/creds/GitHubAppCredential.html)

### Example

Query all branches and tags for jenkinsci/jenkins and resolve latest contributor metadata for each.

https://github.com/samrocketman/jervis/issues/133#issuecomment-1614036278

This resolves the jenkinsci/jenkins Git metadata in under 10 seconds (the GH api library resolves the same information in a couple of hours using over 4000 API requests).

**Comment by bitwiseman** [comment]
We have a first PR accessing a GraphQL mutation - #2056 . 

Still to do: 
* More GraphQL support in Requester or in some kind of query builder. 
* Add GraphQL pagination support to base response and iterables.

**Comment by zhaomoran** [comment]
Following up on the recent work in #2268 (structured GraphQL error handling with `GHGraphQLException`), I'd like to discuss the next step: exposing a public GraphQL query interface.

**Current state:** `GitHub.createGraphQLRequest()` is package-private, so GraphQL is only usable internally (currently just 2 mutations in `GHPullRequest`). This means the library can't leverage GraphQL's core strength — letting callers compose flexible queries for exactly the data they need.

**Problem:** More GitHub features are becoming GraphQL-only (Discussions, Projects v2, Sponsorships, some branch protection rules). The current approach of wrapping each one in a typed domain method doesn't scale, and as @gsmet noted, GraphQL queries really need to be tuned by the caller to minimize cost.

**Proposal:** Expose a controlled public GraphQL interface that reuses the library's auth, rate-limit, and retry infrastructure:

```java
// Option A: raw JSON response
JsonNode result = github.graphql(query, variables);

// Option B: typed response with user-provided POJO
MyResponse result = github.graphql(query, variables, MyResponse.class);
```

This would give callers full query flexibility while the library still manages:
- Authentication (token, GitHub App, etc.)
- Rate limiting (`RateLimitTarget.GRAPHQL`, cost-based)
- Retry and secondary rate limit handling
- Structured error handling (via `GHGraphQLException` from #2268)

**Open questions:**
1. Should pagination (cursor-based) be handled by the library, or left to callers?
2. Return type: Jackson `ObjectNode` vs `Map<String, Object>` vs caller-provided class?
3. Should there be a query builder, or just accept raw query strings (as octokit and @samrocketman's Jervis client do)?

I'm happy to work on a PR for this if there's interest. Would love to hear thoughts from maintainers on the preferred direction.
